cmd/otel-allocator/allocation/least_weighted.go

```diff
@@ -15,291 +15,42 @@
-	The targets are allocated using the least connection method
-	Target Allocator will need information about the collectors in order to set the URLs
-	Keep a Map of what each collector currently holds and update it based on new scrape target updates
-*/
-
-// leastWeightedAllocator makes decisions to distribute work among
-// a number of OpenTelemetry collectors based on the number of targets.
-// Users need to call SetTargets when they have new targets in their
-// clusters and call SetCollectors when the collectors have changed.
-type leastWeightedAllocator struct {
-	// m protects collectors and targetItems for concurrent use.
-	m sync.RWMutex
-	// collectors is a map from a Collector's name to a Collector instance
-	collectors map[string]*Collector
-	// targetItems is a map from a target item's hash to the target items allocated state
-	targetItems map[string]*target.Item
-
-	// collectorKey -> job -> target item hash -> true
-	targetItemsPerJobPerCollector map[string]map[string]map[string]bool
-
-	log logr.Logger
-
-	filter Filter
-}
+var _ Strategy = &leastWeightedStrategy{}
 
-// SetFilter sets the filtering hook to use.
-func (allocator *leastWeightedAllocator) SetFilter(filter Filter) {
-	allocator.filter = filter
-}
+type leastWeightedStrategy struct{}
 
-func (allocator *leastWeightedAllocator) GetTargetsForCollectorAndJob(collector string, job string) []*target.Item {
-	allocator.m.RLock()
-	defer allocator.m.RUnlock()
-	if _, ok := allocator.targetItemsPerJobPerCollector[collector]; !ok {
-		return []*target.Item{}
-	}
-	if _, ok := allocator.targetItemsPerJobPerCollector[collector][job]; !ok {
-		return []*target.Item{}
-	}
-	targetItemsCopy := make([]*target.Item, len(allocator.targetItemsPerJobPerCollector[collector][job]))
-	index := 0
-	for targetHash := range allocator.targetItemsPerJobPerCollector[collector][job] {
-		targetItemsCopy[index] = allocator.targetItems[targetHash]
-		index++
-	}
-	return targetItemsCopy
+func newleastWeightedStrategy() Strategy {
+	return &leastWeightedStrategy{}
 }
 
-// TargetItems returns a shallow copy of the targetItems map.
-func (allocator *leastWeightedAllocator) TargetItems() map[string]*target.Item {
-	allocator.m.RLock()
-	defer allocator.m.RUnlock()
-	targetItemsCopy := make(map[string]*target.Item)
-	for k, v := range allocator.targetItems {
-		targetItemsCopy[k] = v
-	}
-	return targetItemsCopy
+func (s *leastWeightedStrategy) GetName() string {
+	return leastWeightedStrategyName
 }
 
-// Collectors returns a shallow copy of the collectors map.
-func (allocator *leastWeightedAllocator) Collectors() map[string]*Collector {
-	allocator.m.RLock()
-	defer allocator.m.RUnlock()
-	collectorsCopy := make(map[string]*Collector)
-	for k, v := range allocator.collectors {
-		collectorsCopy[k] = v
+func (s *leastWeightedStrategy) GetCollectorForTarget(collectors map[string]*Collector, item *target.Item) (*Collector, error) {
+	// if a collector is already assigned, do nothing
```

Comment:
This is sneaky but perfect for maintaining this functionality 🙂

Reply:
In a follow up, I want to remove the `CollectorName` field from `target.Item` and instead track this in a separate map in the allocator. Then we can pass this map as an argument to this function and be explicit about the dependency.